### PR TITLE
Traffic Control: Alert when a road topic exists without associated model

### PR DESCRIPTION
**Pull request by teivah**
Main change: Creation of a `ModelAssociationChecker` which periodically checks inconsistencies between Kafka and Hive Metastore.

**Please note** this PR is based on https://github.com/HotelsDotCom/data-highway/pull/37 as otherwise we would have used the deprecated Kafka API to list the topics.

### CODE REVIEW

File: agent/traffic-control/README.md [hunk 1]
```diff
@@ -25,19 +25,24 @@ subset that the agent uses with example values:
...
+| `kafka.bootstrapServers`          | -                        | Connection string for Kafka to allow reading of models and writing modifications. In Kubernetes, a typical value would be `kafka:9092`.
+| `kafka.store.topic`               | -                        | Kafka topic for reading models.
+| `kafka.modification.topic`        | -	                       | Kafka topic for sending modifications.
+| `kafka.default.topicConfig`       | -                        | Additional properties to assign to Kakfa topics on creation.
+| `kafka.default.partitions`        | 6                        | The number of partitions to create Kafka topics with.
+| `kafka.default.replicationFactor` | 3                        | The replication factor to create Kafka topics with.
+| `graphite.endpoint`               | disabled                 | Graphite host and port for sending metrics. Disabled by default.
+| `cronScheduler`                   | 0 0 * * * *  (each hour) | Scheduling configuration of the model association checker.
+| `hive.database`                   | -                        | Hive database in which to create tables for landed data.
```
Comment (SandeepSolanki) [inline agent/traffic-control/README.md]: If I understand correctly, the idea here was to check if _roads Model and do the diff against the Kafka Topic for each entry in _roads Model. Reason being if TrafficControl is offline when Road is deleted then the corresponding Kafka topic will not be deleted leading into inconsistency. Checking Hive Table which exists for deleted Topic is not in scope of this Jira. Also currently we are not deleting Hive Data for that Topic.

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import kafka.zk.KafkaZkClient;
+import lombok.extern.slf4j.Slf4j;
+import scala.collection.JavaConversions;
+
+@Component
+@Slf4j
+public class ModelAssociationChecker {
+  public static final String ROAD_PREFIX = "road.";
+
+  private @Autowired KafkaZkClient kafkaZkClient;
```
Comment (nahguam) [inline agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: We usually avoid `@Autowired` fields directly and only autowire constructors. Make it simple to inject for tests.

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  public void run(KafkaZkClient kafkaZkClient, IMetaStoreClient hiveClient, String hiveDb,
+                  Consumer<String> missingTopicStrategy) {
+    final List<String> topics = JavaConversions.seqAsJavaList(kafkaZkClient.getAllTopicsInCluster());
+    topics.stream()
+        .filter(topic -> topic.startsWith(ROAD_PREFIX))
+        .filter(topic -> !isHiveTableExists(hiveClient, hiveDb, topic))
+        .forEach(missingTopicStrategy);
+  }
+
+  public boolean isHiveTableExists(IMetaStoreClient hiveClient, String hiveDb, String tableName) {
```
Comment (nahguam) [inline agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: I may be missing some of the intent behind this but traffic control shouldn't know anything about Hive.

File: agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java [hunk 4]
```diff
@@ -69,7 +70,7 @@ public void run(ApplicationArguments args) throws Exception {
     log.info("Offsets {}", offsets);
     while (running.size() > 0) {
       log.info("Running {}", running.values());
-      for (ConsumerRecord<Void, Record> record : consumer.poll(pollTimeout)) {
+      for (ConsumerRecord<Void, Record> record : consumer.poll(Duration.ofMillis(pollTimeout))) {
```
Comment (nahguam) [inline agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java]: Would be nice if the `pollTimeout` was injected as a `Duration`

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java (hunk 3 above)
Comment (teivah) [reply agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: I thought _have an associated model_ means Hive Metastore. Also according to @Noddy76 :
> So Kafka has a topic `road.foo` but there is nothing in the store for a road called `foo`

I though again we were talking about Hive Metastore...

So, what shall I do?

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java (hunk 2 above)
Comment (teivah) [reply agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: Well, it's still possible with `@InjectMocks` but okay I'll comply with the existing way of doing it.

File: agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java (hunk 4 above)
Comment (teivah) [reply agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java]: Understood but this more a remark for https://github.com/HotelsDotCom/data-highway/pull/37 I guess

File: agent/traffic-control/README.md (hunk 1 above)
Comment (teivah) [reply agent/traffic-control/README.md]: Not sure actually (see https://github.com/HotelsDotCom/data-highway/pull/39#discussion_r258874148)

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java (hunk 3 above)
Comment (nahguam) [reply agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: Ah ok - This means that each Kafka topic should have an associated model in the model topic.

File: agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java (hunk 4 above)
Comment (nahguam) [reply agent/loading-bay/truck-park/core/src/main/java/com/hotels/road/truck/park/TruckPark.java]: yep

File: agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java (hunk 3 above)
Comment (teivah) [reply agent/traffic-control/src/main/java/com/hotels/road/trafficcontrol/ModelAssociationChecker.java]: Should be okay.

File: agent/tollbooth/road-tollbooth-app/src/main/java/com/hotels/road/tollbooth/app/TollboothApp.java [hunk 5]
```diff
@@ -105,15 +105,31 @@ public ObjectMapper objectMapper() {
...
+          MAX_IN_FLIGHT_REQUESTS, Time.SYSTEM, "", "");
+      kafkaZkClient = new KafkaZkClient(zooKeeperClient, IS_SECURE_KAFKA_CLUSTER, Time.SYSTEM);
+      AdminZkClient adminZkClient = new AdminZkClient(kafkaZkClient);
+
+      if (!isTopicExists(topic, adminZkClient)) {
+        adminZkClient.createTopic(topic, partitions, replicas, new Properties(), RackAwareMode.Enforced$.MODULE$);
+      }
+    } finally {
+      if (kafkaZkClient != null) {
+        kafkaZkClient.close();
```
Comment (rshotels) [inline agent/tollbooth/road-tollbooth-app/src/main/java/com/hotels/road/tollbooth/app/TollboothApp.java]: can this throw an Exception?

**Comment by nahguam** [comment]
I'm going to review this after the cluster is upgraded.

**Pull request closed without merging**